src/qe/mbp/mbp_term_graph.cpp

```diff
@@ -445,130 +740,355 @@ namespace mbp {
-    void term_graph::pick_roots() {
-        SASSERT(marks_are_clear());
-        for (term* t : m_terms) {
-            if (!t->is_marked() && t->is_root())
-                pick_root(*t);
-        }
-        reset_marks();
+    return true;
+  }
+
+  //pick representatives for all terms in todo. Then, pick representatives for
+  //all terms whose children have representatives
+  void term_graph::pick_repr_percolate_up(ptr_vector<term>& todo) {
+    term* t;
+    while(!todo.empty()) {
+      t = todo.back();
+      todo.pop_back();
+      if (t->get_repr()) continue;
+      pick_repr_class(t);
+      for (auto it : term::parents(t->get_root()))
+        if (all_children_picked(it)) todo.push_back(it);
+    }
+  }
+
+  //iterate through all terms in a class and pick a representative that:
+  // 1. is cgr and 2. least according to term_lt
+  void term_graph::pick_repr_class(term *t) {
+    SASSERT(all_children_picked(t));
+    term *r = t;
+    for (term *it = &t->get_next(); it != t; it = &it->get_next()) {
+      if (!all_children_picked(it)) continue;
+      if ((it->is_cgr() && !r->is_cgr()) ||
+          (it->is_cgr() == r->is_cgr() && term_lt(*it, *r)))
+        r = it;
     }
+    r->mk_repr();
+  }
+
+  // Choose repr for equivalence classes
+  // repr has the following properties:
+  // 1. acyclicity (mk_app terminates)
+  // 2. maximal wrt cgr
+  // 3. each class has exactly one repr
+  // assumes that cgroundness has been computed
+  void term_graph::pick_repr() {
+    //invalidates cache
+    m_term2app.reset();
+    DEBUG_CODE(for (term* t : m_terms) SASSERT(t->deg() == 0 || !t->all_children_ground() || t->is_cgr()););
+    for (term* t : m_terms) t->reset_repr();
+    ptr_vector<term> todo;
+    for (term *t : m_terms) {
+      if (t->deg() == 0 && t->is_cgr())
+        todo.push_back(t);
+    }
+    pick_repr_percolate_up(todo);
+    DEBUG_CODE(for (term* t : m_terms) SASSERT(!t->is_cgr() || t->get_repr()););
 
-    void term_graph::display(std::ostream &out) {
-        for (term * t : m_terms) {
-            out << *t;
-        }
+    for (term *t : m_terms) {
+      if (t->get_repr()) continue;
+      if (t->deg() == 0)
+        todo.push_back(t);
+    }
+    pick_repr_percolate_up(todo);
+    DEBUG_CODE(for (term* t : m_terms) SASSERT(t->get_repr()););
+    DEBUG_CODE(for(auto t : m_terms) SASSERT(!t->is_cgr() || t->get_repr()->is_cgr()););
+  }
+
+  // if t is a variable, attempt to pick non-var
+  void term_graph::refine_repr_class(term* t) {
+    SASSERT(t->is_repr());
+    auto is_var = [&] (term *p) {
+      SASSERT(is_app(p->get_expr()));
+      return m_is_var.contains(to_app(p->get_expr())->get_decl());
+    };
+    if (!is_var(t)) return;
+    term *r = t;
+    for (term *it = &t->get_next(); it != t; it = &it->get_next()) {
+      if (makes_cycle(it)) continue;
+      if (is_var(r) && !is_var(it))
+        r = it;
     }
+    r->mk_repr();
+  }
+
+  // check if t makes a cycle if chosen as repr
```

Comment:
Additional comment: This function assumes that the current `repr` doesn't have cycles. If there is a cycle in a child class it doesn't terminate.